### ISSUE TITLE
Support Azure Open AI

### ISSUE THREAD

**Issue by pantherale0** [valentinfrlch/ha-llmvision#64]
**Is your feature request related to a problem? Please describe.**
Currently this integration does not support Azure OpenAI which is works slightly differently to a standard OpenAI custom server. 

Integration setup fails with `Could not connect to the server. Check you API key or IP and port`

**Describe the solution you'd like**
Support for Azure OpenAI endpoints.

**Describe alternatives you've considered**
Using a proxy to proxy the requests, however this solution is not as clean as a native solution.

**Additional context**
N/A

**Comment by valentinfrlch** [comment]
Will look into this

**Comment by valentinfrlch** [comment]
Have you tried using the custom OpenAI compatible option?

**Comment by pantherale0** [comment]
Hi,

Yes, however it doesn't work with the above error logged. Example endpoint URL:

`https://{resource_name}.openai.azure.com/openai/deployments/gpt-4o/chat/completions?api-version=2023-03-15-preview`

**Comment by valentinfrlch** [comment]
Ah sorry you wrote that initially. Is that still the case? I updated the custom OpenAI provider in the last update.

**Comment by pantherale0** [comment]
No problem, yes, its still the case, tried it again today.

**Comment by xiasi0** [comment]
#86 Our situation seems to be the same. Have you solved it yet?

**Comment by xiasi0** [comment]
> I believe the API itself is actually slightly different to the standard OpenAI spec which is why it doesn't work. But #86 will most likely use the same service as the PAYG production instances (that I'm trying to use).

I forcibly modified the BASE URL address for OpenAI Conversation core integration, using url: https://models.inference.ai.azure.com It is available

But LLM Vision cannot.😂

**Comment by valentinfrlch** [comment]
Looking at [Azure OpenAI REST API Reference](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference) it is pretty obvious why it doesn't work. Azure OpenAI is not compatible with OpenAI's API.
The endpoints are somewhat similar (`/chat/completions` vs `/{deployment_id}/completions`), but the actual request body is completely different.

**OpenAI**:
```
{
     "model": "gpt-4o-mini",
     "messages": [{"role": "user", "content": "Say this is a test!"}],
     "temperature": 0.7
}
```

**Azure**:
```
{
 "prompt": [
  "tell me a joke about mango"
 ],
 "max_tokens": 32,
 "temperature": 1.0,
 "n": 1
}
```

For now I can only recommend using OpenAI directly instead, or if you're looking for a free provider, I believe Gemini has a free tier.

**Comment by xiasi0** [comment]
> Looking at [Azure OpenAI REST API Reference](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference) it is pretty obvious why it doesn't work. Azure OpenAI is not compatible with OpenAI's API. The endpoints are somewhat similar ( vs ), but the actual request body is completely different.`/chat/completions``/{deployment_id}/completions`
> 
> **OpenAI**:
> 
> ```
> {
>      "model": "gpt-4o-mini",
>      "messages": [{"role": "user", "content": "Say this is a test!"}],
>      "temperature": 0.7
> }
> ```
> 
> **Azure**:
> 
> ```
> {
>  "prompt": [
>   "tell me a joke about mango"
>  ],
>  "max_tokens": 32,
>  "temperature": 1.0,
>  "n": 1
> }
> ```
> 
> For now I can only recommend using OpenAI directly instead, or if you're looking for a free provider, I believe Gemini has a free tier.

If it's worth the effort. I hope it can be compatible because Gemini is not available in my region. My GPU no longer has any more VRAM, so I cannot use LLM Vision locally. I cannot afford more bills and more VRAM GPUs.

**Comment by pantherale0** [comment]
@valentinfrlch is there a reason OpenAI's python package isn't used to facilitate communication to OpenAI based LLMs?

**Comment by valentinfrlch** [comment]
I figured, since there are some other providers I might as well just use http requests for all providers as this is easier to maintain. I am in the process of rewriting the request_helpers.py though.

Is there a reason why the python package should be used?

**Comment by kiloptero** [comment]
same here.. Im not able to use the azure open ai.. Will be great to be Able to use it.

**Comment by valentinfrlch** [comment]
It wasn't added in v1.4.0. If someone would like to finish the implementation for Azure, PRs are always welcome!

**Comment by Anton2079** [comment]
What’s required, I might be able to assist.

**Comment by valentinfrlch** [comment]
The implementation is already there, but needs to be tested. Some changes might also be necessary because I don't think it works yet. I haven't tested this.

This is where the request logic is handled:
https://github.com/valentinfrlch/ha-llmvision/blob/8406ea4938c0b782c9a4af250086917f22ae4a8e/custom_components/llmvision/providers.py#L446C1-L518C67

This is where the provider class is initialized:
https://github.com/valentinfrlch/ha-llmvision/blob/8406ea4938c0b782c9a4af250086917f22ae4a8e/custom_components/llmvision/providers.py#L151C9-L164C47

And finally, this is where setup is handled:
https://github.com/valentinfrlch/ha-llmvision/blob/8406ea4938c0b782c9a4af250086917f22ae4a8e/custom_components/llmvision/config_flow.py#L301C5-L349C10

Thank you for offering your help! If you need any help, just let me know!

**Comment by PavolC** [comment]
I have been running the code as is with Azure enabled in my instance for the last couple months and it works great!

**Comment by mrahmadt** [comment]
@PavolC can you please explain how did you manage to get it to work? I'm still getting this error

Could not connect to the server. Check you API key or IP and port

**Comment by lluisd** [comment]
> I have been running the code as is with Azure enabled in my instance for the last couple months and it works great!

so, you only did these manual changes of these 3 files to make it work? or something else? I will also want to test it.

**Comment by PavolC** [comment]
Only thing I did was add 'Azure' to the list here: https://github.com/valentinfrlch/ha-llmvision/blob/8406ea4938c0b782c9a4af250086917f22ae4a8e/custom_components/llmvision/config_flow.py#L92

This enabled Azure from the dropdown when configuring the integration.

@mrahmadt sounds maybe like the values you provided for the Azure authentication settings might not be correct? Assuming you have a model deployed into Azure AI Foundry?

**Comment by valentinfrlch** [comment]
@PavolC Thank you for reporting back! What version are you running? If this is working, we can merge it into 1.5.1.

**Comment by PavolC** [comment]
@valentinfrlch I'm using 1.5.0

**Comment by lluisd** [comment]
Azure openai working fine also for me.

**Comment by kiloptero** [comment]
same for me

**Comment by valentinfrlch** [comment]
@lluisd @kiloptero Did you try with the v1.5.1 Beta 2?

**Comment by lluisd** [comment]
> [@lluisd](https://github.com/lluisd) [@kiloptero](https://github.com/kiloptero) Did you try with the v1.5.1 Beta 2?

I was using 1.5.0 but now I upgraded to v1.5.1 Beta 2 and it's also working fine. 

One non related question at all: why on reconfigure the Azure provider it doesn't show the current configuration to edit? I don't see a way to see what I configured.

**Comment by valentinfrlch** [comment]
@lluisd Thanks for the feedback. The issue you've encountered is actually not unrelated, Azure didn't get the updated provider settings from v1.5.0 (as it wasn't enabled then).

**Update:** Should be fixed in Beta 3. Full release notes: https://github.com/valentinfrlch/ha-llmvision/releases/tag/v1.5.1-beta.3

**Comment by lluisd** [comment]
Yes, now I can see all the fields that I filled on reconfiguring the Azure provider.

**Comment by valentinfrlch** [comment]
We've added support for Azure in v1.5.1. If there are any problems with it, please open a new issue.
